### PR TITLE
Rest api client

### PR DESCRIPTION
**Pull request by Lupino**
- add a rest api client library
- add rest api hotgym example

### CODE REVIEW

File: py/htm/examples/rest/client.py [hunk 1]
```diff
@@ -27,11 +27,11 @@
 #        port default: 8050
 #
 
-import requests
 import sys
-import json
 import math
 
+import rest_api
```
Comment (breznak) [inline py/htm/examples/rest/client.py:33]: if we use this, I'd suggest more explicit name. This looks like a generic 3rd-party REST library - `htm_rest_api`?

File: py/htm/examples/rest/client.py [hunk 2]
```diff
@@ -53,20 +53,11 @@ def main(argv):
        {addLink:   {src: "encoder.encoded", dest: "sp.bottomUpIn"}},
        {addLink:   {src: "sp.bottomUpOut", dest: "tm.bottomUpIn"}}
     ]}"""
+
+  net = rest_api.Network(config, host = URL, verbose=verbose)
     
   # Send the config string and obtain a token for the created Network object instance.
-  if verbose: print('Post /network; body: '+config)
-  res = requests.post(URL+'/network', data=config)
```
Comment (breznak) [inline py/htm/examples/rest/client.py:59]: I like the simplification this provides, but I don't know if the example shouldn't be explicit on how the REST is being used here

File: py/htm/examples/rest/hotgym.py [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+from rest_api import Link, Region, NetworkConfig, Network, get_classifer_predict
+
+_EXAMPLE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+_INPUT_FILE_PATH = os.path.join(_EXAMPLE_DIR, "gymdata.csv")
+
+default_parameters = {
+    # there are 2 (3) encoders: "value" (RDSE) & "time" (DateTime weekend, timeOfDay)
+    'enc': {
+        "value": {
+            'resolution': 0.88,
```
Comment (breznak) [inline py/htm/examples/rest/hotgym.py:16]: I haven't checked, but these params should match those of "pythonic hotgym.py" example.

File: py/htm/examples/rest/rest_api.py [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+
+  if idx == -1:
+    return text
+
+  data = text[idx + 1:].replace(']}', '')
+
+  return json.loads('[' + data + ']')
+
+
+class Network(object):
```
Comment (breznak) [inline py/htm/examples/rest/rest_api.py:32]: this helper class indeed simplifies the usage of the REST API significantly :+1: If it's generic enough for use, we should place it in a proper place - `/bindings/rest/`?
CC @dkeeney please weight in

File: src/htm/regions/SPRegion.cpp [hunk 5]
```diff
@@ -43,7 +43,7 @@ SPRegion::SPRegion(const ValueMap &values, Region *region)
   // parameters out of the map and set aside so we can pass them to the SpatialPooler
   // algorithm when we create it during initialization().
   args_.columnCount = values.getScalarT<UInt32>("columnCount", 0);
-  args_.potentialRadius = values.getScalarT<UInt32>("potentialRadius", 16u);
+  args_.potentialRadius = values.getScalarT<UInt32>("potentialRadius", 0u);
```
Comment (breznak) [inline src/htm/regions/SPRegion.cpp:46]: why the change? if it matches the default in SP, then OK.

File: py/htm/examples/rest/client.py (hunk 2 above)
Comment (breznak) [reply py/htm/examples/rest/client.py:59]: ..depends on our stance below https://github.com/htm-community/htm.core/pull/834#discussion_r437267218

**Comment by breznak** [comment]
@Lupino some tests failing 
https://github.com/htm-community/htm.core/pull/834/checks?check_run_id=752256705#step:8:1091 

Some might be random in ClassifierRegion, one I suspect is caused by the SPRegion's param change.

**Comment by Lupino** [comment]
> @Lupino some tests failing
> https://github.com/htm-community/htm.core/pull/834/checks?check_run_id=752256705#step:8:1091
> 
> Some might be random in ClassifierRegion, one I suspect is caused by the SPRegion's param change.

I found the document say  potentialRadius default value is zero

File: py/htm/examples/rest/hotgym.py (hunk 3 above)
Comment (Lupino) [reply py/htm/examples/rest/hotgym.py:16]: > I haven't checked, but these params should match those of "pythonic hotgym.py" example.

yes i copy from then py/htm/examples/hotgym.py

**Comment by dkeeney** [comment]
> where is the best place to put the htm_rest_api file? I'd start a bindings/rest/ (?)

That is a good place to put it.  
This module is dependent on the server which is in the example code. src/examples/rest/server.core.hpp.   If we bring the server into the main body of htm.core it should probably not be in the examples folder.  What do you think?

> will we want a similar NetworkREST file for the c++, or the "raw REST" is fine there?

Well...I guess we could.  It would not be part of the htm.core library because the whole point of the REST API is to avoid the requirement that the client side needs to load the htm.core library.

There is also  the copywrite to consider. One of the reasons to have the REST interface is to separate code in the Numenta copyright from someone else's intellectual property.  So, if you require anything from our library to be included in the client side then it could violate a copyright.  That is why I put the client stuff in the examples.  Something to think about.

**Comment by breznak** [comment]
> If we bring the server into the main body of htm.core it should probably not be in the examples folder.  

good point. yes, move (minimum) what is needed to the new `xxx/rest/` location. 

 >  It [NetworkREST c++] would not be part of the htm.core library because the whole point of the REST API is to avoid the requirement that the client side needs to load the htm.core library. 

good catch! 
On the other hand, this new NetworkREST does not include any of `htm.*`,right? So while you still obviously need to know what you're doing (~know HTM NAPI), you don't need to have htm.core installed (ie it's not available for your platform)
https://github.com/htm-community/htm.core/pull/834/files#diff-563b1ac3435bfc941454f6a46de67913R1-R6 

So this could be a very well a header-only binding (c++,py) which you CAN copy/download from our repo and use in your project. 

> There is also the copywrite to consider. One of the reasons to have the REST interface is to separate code in the Numenta copyright from someone else's intellectual property. So, if you require anything from our library to be included in the client side then it could violate a copyright.  

- you could (probably?) licence your class/code (not depending on HTMcore) with whatever licenc (say BSD). 
- I don't think we should (maybe we could) do that. [Licences in my layman terms: GPL: "we want this code to remain open-source, if you distribute it for commercial purposes, you have to contribute back you changes for others to benefit". This was to fight 'tivoization'. Then with the movement 'everything runs in cloud', this started to be abused, as a company did not distribute any binaries, the (OSS) service ran in the cloud. *A*GPL (ours licence) covers that: "if you USE OSS SW, you have to contribute back". I think this covers the usecase  REST communicates with a HTM server.

**Comment by breznak** [comment]
> So, if you require anything from our library to be included in the client side then it could violate a copyright. That is why I put the client stuff in the examples 

If you say the REST py depends on REST c++ (rest_server.hpp) then that includes other HTM NAPI, so it is a derived work imho, and the licence is the same as ours, AGPL. 

https://github.com/htm-community/htm.core/blob/master/src/examples/rest/server_core.hpp#L62

@Lupino could you update the `src/examples/rest/README` with the new info about this PY code, please?

**Comment by dkeeney** [comment]
> so it is a derived work imho, and the licence is the same as ours, AGPL.

I think that is the point.  I would like it to NOT be derived work.  No more than if they called a function in the OS.  I don't want to get into a long discussion of licensing but I think we need to be clear among ourselves what we want covered.  

But perhaps there is no way around it. It is covered under AGPL even if we provide no client code.

Oh, also note that the REST server we use is from a third party, and has an MIT license
```
//  Copyright (c) 2019 Yuji Hirose. All rights reserved.
//  MIT License
```
That should probably be listed someplace in the README.md

File: py/tests/rest/htm_rest_api_test.py [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+    r = net.delete_all()
+    self.assertEqual(r, 'OK')
+
+  def tearDown(self):
+    try:
+      r = requests.get('{}/stop'.format(HOST))
+      print(r)
+    except:
+      pass
+    self._thread.join()
```
Comment (dkeeney) [inline py/tests/rest/htm_rest_api_test.py:204]: I think it is hanging here.
Is there a way to kill another thread in Python?
If you should get an exception while sending the "/stop" message or the server does not handle it for any reason, the thread.join( ) would wait forever.

**Comment by dkeeney** [comment]
I could have this same shutdown problem with my C++ version of the test. The server does not shut down if the /stop command is not processed.

I was thinking about some sort of server timeout after the last request as a fail-safe measure but still looking into this.

File: py/tests/rest/htm_rest_api_test.py [hunk 7]
```diff
@@ -200,7 +200,7 @@ def tearDown(self):
       r = requests.get('{}/stop'.format(HOST))
       print(r)
     except:
-      pass
+      os.system("ps x | grep rest_server | grep -v grep | awk '{print $1}' | xargs kill -9")
```
Comment (breznak) [inline py/tests/rest/htm_rest_api_test.py:203]: I guess this will only run on Unixes, not Windows. On the other hand, I wouldn't worry too much about stopping the server? Since the intended use is that the user runs manually the server (so they should as well bring it down). Ofc this is a problem for the automated test - wouldn't the OS kill all threads spawned from the (unit-test) process once the parent finishes?

Comment (Lupino) [reply py/tests/rest/htm_rest_api_test.py]: I' ll remove it.

Comment (dkeeney) [reply py/tests/rest/htm_rest_api_test.py]: > wouldn't the OS kill all threads spawned from the (unit-test) process once the parent finishes?

Good point.  For the unit test we could just remove the line `threadObj.join()`.  
If the server does not process the /stop then it will be killed when the test program finishes.  No problem with unknown state or anything.  Everything is clean.

In a general case where the server is in a different machine we would not want to be shutting down the server.  So this is good.

Thanks.

File: py/tests/rest/htm_rest_api_test.py (hunk 6 above)
Comment (dkeeney) [reply py/tests/rest/htm_rest_api_test.py:200]: Oh, you need to remove the line "self._thread.join( )."

Comment (breznak) [reply py/tests/rest/htm_rest_api_test.py]: should be good now?

**Comment by dkeeney** [comment]
Hmmm, on OSx it still seems to hang.

**Comment by dkeeney** [comment]
I am not sure about Python but it appears that letting the main thread terminate before all child threads have terminated is a bad thing in C++.  The main thread should either join the child threads, i.e. wait for the other thread to end, or detach the other threads before terminating.

I am struggling with this in PR #837 where the C++ version of this is also hanging if I join or crashing if I let the main thread terminate while the server thread remains running.

Don't have a solution yet.

**Comment by dkeeney** [comment]
@Lupino for contributing.

**Comment by breznak** [comment]
Thank you for the contribution and bearing with the PR process :) :+1:
